docs/configuration.md
### dev-lxc.yml Config Files

dev-lxc uses a YAML configuration file named `dev-lxc.yml` to define a cluster.

The `init` command generates sample config files for various server types. You can run `dl help init` to see the many types of server configurations that can be generated.

The `--base-container` option can be used to set the type of base container you want used in the generated configuration.

The `--product-versions` option can be used to set the package channel and version to use for multiple products. If a version is not specified then it will default to `latest`. If `none` is used then that product will be removed from the generated configuration.

For example, the following command will generate a configuration for a standalone Chef Server that uses the `b-centos-7` base container and install chef-server 12.13.0 from the `stable` channel, the latest version of manage from the `current` channel and remove the reporting product.

```
dl init --chef --base-container b-centos-7 --product-versions chef-server:stable:12.13.0 manage:current reporting:none
```

Let's generate a config for a cluster with a standalone Chef Server, Supermarket server,
Compliance server, Chef Automate server and a Job Dispatch Runner and remove the "reporting" product from the generated configuration.

```
dl init --chef --compliance --supermarket --automate --runners --product-versions reporting:none > dev-lxc.yml
```

We can easily append additional configurations to this file. For example, the following command appends an infrastructure node.

```
dl init --nodes -a >> /root/work/clusters/automate/dev-lxc.yml
```

The contents of `dev-lxc.yml` should look like this.

```
# enable_build_snapshots automatically makes container snapshots at key times during the build process
# default value is `true`
#enable_build_snapshots: true

# base_container must be the name of an existing container
base_container: b-ubuntu-1404

# memory_per_server sets the maximum amount of user memory (including file cache) for each server.
# dev-lxc will set the `memory.limit_in_bytes` cgroup for each server to apply this limit.
# If no units are specified, the value is interpreted as bytes.
# You can use suffixes to represent larger units - k or K for kilobytes, m or M for megabytes, and g or G for gigabytes.
# The default behavior is that no limit is set.
#memory_per_server: 4G

# list any host directories you want mounted into the servers
#mounts:
#  - /root/clusters root/clusters

# list any SSH public keys you want added to /home/dev-lxc/.ssh/authorized_keys
#ssh-keys:
#  - /root/clusters/id_rsa.pub

# DHCP reserved (static) IPs must be selected from the IP range 10.0.3.150 - 254

chef-server:
  users:          # a user's password will be the same as its username
    - mary-admin
    - joe-user
  orgs:
    demo:
      admins:
        - mary-admin
      non-admins:
        - joe-user
  servers:
    chef.lxc:
      ipaddress: 10.0.3.203
      products:
        chef-server:
          channel: stable
          version: latest
        manage:
          channel: stable
          version: latest
        push-jobs-server:
          channel: stable
          version: latest

compliance:
  admin_user: admin         # the password will be the same as the username
  servers:
    compliance.lxc:
      ipaddress: 10.0.3.205
      products:
        compliance:
          channel: stable
          version: latest

supermarket:
  servers:
    supermarket.lxc:
      ipaddress: 10.0.3.206
      products:
        supermarket:
          channel: stable
          version: latest

automate:
  servers:
    automate.lxc:
      ipaddress: 10.0.3.200
      products:
        automate:
          channel: stable
          version: latest
      license_path: ../delivery.license
      chef_org: automate
      enterprise_name: default

runners:
  servers:
    runner-1.lxc:
      products:
        chefdk:     # downloaded only
          channel: stable
          version: latest

nodes:
  chef_server_url: https://chef.lxc/organizations/demo
  validation_client_name: demo-validator
  # comment out or remove the validation_key path to use chef-server keys generated by dev-lxc
  validation_key: # /path/for/ORG-validator.pem
  servers:
    node-1.lxc:
      products:
        chef:
          channel: stable
          version: latest

```

The dev-lxc.yml config file is very customizable. You can add or remove mounts, products or servers,
change ip addresses, server names, the base_container and more.

As you can see there are six server types represented by six servers.

1. chef-server - chef.lxc
2. compliance - compliance.lxc
3. supermarket - supermarket.lxc
4. automate - automate.lxc
5. runners - runner-1.lxc
6. nodes - node-1.lxc

#### Global Settings

The global settings used by each of the server types are `enable_build_snapshots`, the `base_container`, `memory_per_server`, a list of `mounts` and a list of `ssh-keys`. These settings are described in the config comments.

Be sure to set `base_container` in the `dev-lxc.yml` to an existing container's name.  
This container will be cloned to create each container in the cluster.  
If you don't already have a container to use as a `base_container` then you can follow the instructions in the  
[Create a dev-lxc Base Container section](https://github.com/jeremiahsnapp/dev-lxc#create-a-dev-lxc-base-container) to create one.

#### Server Specific Settings

It is possible to define different values for `enable_build_snapshots`, `base_container`, `memory_per_server`, `mounts` or `ssh-keys` for a particular server type or even for a particular server as you can see in the following snippet.

```
nodes:
  base_container: b-ubuntu-1404
  servers:
    node-1.lxc:
      base_container: b-centos-6
    node-2.lxc:
```

IP addresses from the range 10.0.3.150 - 254 can be assigned to the servers. If an IP address
is not specified then a dynamic IP address is assigned when the server starts.

#### mixlib-install Library Automatically Manages a Cache of Product Packages

dev-lxc uses the [mixlib-install](https://github.com/chef/mixlib-install) library to download Chef products
to a cache in `/var/dev-lxc` in the host instance. This cache is automatically mounted into each server when it starts.

A list of Chef products to be installed can be defined for each server
using [product names that mixlib-install understands](https://github.com/chef/mixlib-install/blob/master/PRODUCT_MATRIX.md).

The channel and version of the product can be defined also.

Channel can be `current`, `stable` or `unstable` with `stable` as the default.
Version can be `latest` or a version number with `latest` as the default.

For example, the following specifies the `current` channel and version `0.16.1` of the `chefdk` product.

```
nodes:
  servers:
    node-1.lxc:
      products:
        chefdk:
          channel: current
          version: 0.16.1
```

The `package_source` setting can be used to specify a package file on disk.

```
nodes:
  servers:
    node-1.lxc:
      products:
        chefdk:
          package_source: /root/chefdk_0.16.1-1_amd64.deb
```

#### Chef Server Orgs/Users

When defining a Chef Server you can include organizations and users that will be automatically created and associated accordingly.

#### Specify Chef Server Predefined Full and Partial Configurations

The following options work for a standalone Chef Server, the backend and frontends of a tier Chef Server cluster and the frontends of a chef-backend HA Chef Server cluster.

The `chef-server.rb` option can be used to fill a Chef server's chef-server.rb with predefined content from a file.

The `chef-server.rb_partials` option can be used to append predefined content from multiple files to a Chef server's chef-server.rb file.

```
servers:
  chef.lxc:
    chef-server.rb: /path/to/a/full/configuration/file
    chef-server.rb_partials:
      - /path/to/a/partial/configuration/file
      - /path/to/another/partial/configuration/file
    ipaddress: 10.0.3.203
    products:
      chef-server:
        channel: stable
        version: latest
```

#### Specify Automate Server Predefined Full and Partial Configurations

The `delivery.rb` option can be used to fill an Automate server's delivery.rb with predefined content from a file.

The `delivery.rb_partials` option can be used to append predefined content from multiple files to an Automate server's delivery.rb file.

```
servers:
  automate.lxc:
    delivery.rb: /path/to/a/full/configuration/file
    delivery.rb_partials:
      - /path/to/a/partial/configuration/file
      - /path/to/another/partial/configuration/file
    ipaddress: 10.0.3.200
    products:
      automate:
        channel: stable
        version: latest
```

#### Automatic Integration Between Servers

dev-lxc knows how to automatically configure Chef Server standalone, Chef Server tier topology,
Chef Server HA 2.0 as well as Chef Automate, Chef Client, Analytics, Compliance and Supermarket.

If a Chef Automate server, Compliance server, Analytics server or Supermarket server is defined in the same config file as
a Chef Server then each server will automatically be integrated with that Chef Server.

If a node server with Chef Client or Chef DK installed is defined in the same config file as

CHANGELOG.md
# dev-lxc Change Log

## 3.3.1 (2017-04-13)

* Change Automate's default chef_org and enterprise_name

## 3.3.0 (2017-03-13)

* Increase "waiting for container's network" timeout to 60s
* Provide warning if no servers are defined in dev-lxc.yml
* Fix the way additional frontends in a chef-backend cluster are configured
* Minor improvement to how a chef-backend's frontend chef-server.rb is generated
* Add delivery.rb and delivery.rb_partials config options
* Add chef-server.rb and chef-server.rb_partials config options
* Restart systemd-resolved.service every time a container starts

## 3.2.0 (2017-03-10)

* Add --base-container and --product-versions options to the `init` subcommand

## 3.1.0 (2017-03-08)

* Add example dev-lxc.yml files
* Change product name from 'delivery' to 'automate'
* Minor cleanup in adhoc_clusters docs
* Move byobu keybindings docs to its own file
* Minor improvements to README
* Minor updates about how to use tinyproxy or mitmproxy

## 3.0.0 (2017-03-06)

* Rename `dev-lxc` binary to `dl`
* Comment out the "reporting" product in generated config files since it is a legacy product.
* Change default automate license_path in generated config files to something more readily usable.
* Change default mounts and ssh-keys paths in generated config files
* Overhaul documentation

## 2.7.0 (2017-03-03)

* Make chef-repo command only create .chef directory

## 2.6.2 (2017-02-21)

* Replace unicode dash with ascii hyphen

## 2.6.1 (2017-02-10)

* Fix http_proxy and https_proxy in generated knife.rb and client.rb files

## 2.6.0 (2017-02-09)

* Make api_fqdn unique for chef-tier and chef-backend configs
* Add https_proxy to generated knife.rb and client.rb files

## 2.5.0 (2017-02-08)

* Add memory_per_server config option to limit memory per server

## 2.4.0 (2017-02-03)

* Refactor DHCP address management
* Add enable_build_snapshots config option
* Skip abort test if server exists AND "build: completed" snapshot exists
* Add auto-integration of Compliance and standalone Chef Server
* Remove period from sha256sum error message
* Reword message when base container already exists
* Update prerequisites.md

## 2.3.3 (2016-12-19)

* Remove debug output from "dpkg -i"

## 2.3.2 (2016-12-02)

* Fix DevLXC::Cluster#configure_automate so the reconfigure comes after all changes to delivery.rb
* Add ability to save run_command output to a file
* Use delivery-ctl create-enterprise
* Replace run_ctl with simpler run_command
* Use installed_products instead of required_products in DevLXC::Cluster#configure_products
* Hard stop the server when destroying it
* Fix destroy so confirmation message works

## 2.3.1 (2016-11-28)

* Create DevLXC::Cluster.abort_up to help with code clarity
* Shutdown container gracefully instead of doing a hard stop
* Skip container start if it is already running
* Refactor DevLXC.assign_ip_address to DevLXC::Server.assign_static_ip_address
* Change DevLXC::Server.deregister_from_dhcp to DevLXC::Server.remove_static_ip_address
* Remove dhcp release post-stop hook since it is not the most reliable solution
* Remove lingering DHCP IP addresses before starting a server
* Remove lingering DHCP IP addresses after destroying or shutting down a server
* Comment out data_collector config in a node's client.rb
* Add a clarifying comment to DevLXC::Cluster.up

## 2.3.0 (2016-11-22)

* Create DNS records for all servers before starting a server
* Delete DNS records for all servers if all are stopped
* Add ability to build Automate Job Dispatch Runners
* Enable Compliance profiles asset store in Automate
* Enable Automate Compliance profiles proxy in Chef Server
* Add ability to setup Compliance and create admin user
* Fix dev-lxc.yml parsing bug
* Allow for snapshot comments to be empty
* Fix 'up' command's output so there is a blank line between each server

## 2.2.7 (2016-09-26)

* Improve file existence test during dev-lxc.yml validation

## 2.2.6 (2016-09-26)

* Change default path for ORG-validator.pem for clarity
* Change default path for automate.license for clarity

## 2.2.5 (2016-08-24)

* Print full product cache path in SHA256 error message

## 2.2.4 (2016-08-24)

* Add sha256 check for downloaded packages

## 2.2.3 (2016-07-27)

* Copy validation key when validation_key is set

## 2.2.2 (2016-07-20)

* Don't calculate/install required products if "build: completed" snapshot exists
* Be more explicit about skipping install of chefdk on build nodes

## 2.2.1 (2016-07-20)

* Add --include-products option to show-config command

## 2.2.0 (2016-07-20)

* Make build nodes reregister to Chef Server

lib/dev-lxc/cli.rb
require 'erb'
require "yaml"
require 'dev-lxc'
require 'thor'

module DevLXC::CLI
  class DevLXC < Thor

    no_commands{
      def get_cluster(config_file=nil)
        config_file ||= "dev-lxc.yml"
        if ! File.exists?(config_file)
          puts "ERROR: Cluster config file '#{config_file}' does not exist."
          puts "       Create a `./dev-lxc.yml` file or specify the path using `--config`."
          exit 1
        end
        begin
          cluster_config = YAML.load(IO.read(config_file))
        rescue Psych::SyntaxError => e
          puts "ERROR: A YAML syntax error was found at line #{e.line} column #{e.column}"
          exit 1
        end
        ::DevLXC::Cluster.new(cluster_config)
      end

      def print_elapsed_time(elapsed_time)
        printf "dev-lxc is finished. (%im %.2fs)\n", elapsed_time / 60, elapsed_time % 60
      end
    }

    desc "show-config", "Show calculated configuration"
    option :config, :desc => "Specify a cluster's YAML config file. `./dev-lxc.yml` will be used by default"
    option :include_products, :type => :boolean, :desc => "Calculate required products"
    def show_config
      get_cluster(options[:config]).show_config(options[:include_products])
    end

    desc "create-base-container [BASE_CONTAINER_NAME]", "Create a base container"
    option :options, :aliases => "-o", :desc => "Specify additional options for the lxc create"
    def create_base_container(base_container_name=nil)
      start_time = Time.now
      base_container_names = %w(b-ubuntu-1204 b-ubuntu-1404 b-ubuntu-1604 b-centos-5 b-centos-6 b-centos-7)
      if base_container_name.nil? || ! base_container_names.include?(base_container_name)
        base_container_names_with_index = base_container_names.map.with_index{ |a, i| [i+1, *a]}
        print_table base_container_names_with_index
        selection = ask("Which base container do you want to create?", :limited_to => base_container_names_with_index.map{|c| c[0].to_s})
        base_container_name = base_container_names[selection.to_i - 1]
      end
      ::DevLXC.create_base_container(base_container_name, options[:options])
      puts
      print_elapsed_time(Time.now - start_time)
    end

    desc "init", "Provide a cluster config file"
    option :chef, :type => :boolean, :desc => "Standalone Chef Server"
    option :chef_tier, :type => :boolean, :desc => "Chef Server using Tier topology with one backend"
    option :chef_backend, :type => :boolean, :desc => "Chef Server using Chef Backend HA topology with three backends"
    option :nodes, :type => :boolean, :desc => "Node Servers"
    option :analytics, :type => :boolean, :desc => "Analytics Server"
    option :compliance, :type => :boolean, :desc => "Compliance Server"
    option :supermarket, :type => :boolean, :desc => "Supermarket Server"
    option :automate, :type => :boolean, :desc => "Automate Server"
    option :build_nodes, :type => :boolean, :desc => "Build Nodes"
    option :runners, :type => :boolean, :desc => "Runners"
    option :adhoc, :type => :boolean, :desc => "Adhoc Servers"
    option :base_container, :default => 'b-ubuntu-1404', :desc => "Specify the base container name"
    option :product_versions, :type => :hash, :default => {}, :desc => "Specify channel and version (default 'latest') for products [product:channel,version ...]; [product:none] removes the product from the config"
    option :append, :aliases => "-a", :type => :boolean, :desc => "Do not generate the global config header"
    option :filename, :aliases => "-f", :desc => "Write generated content to FILE rather than standard output."
    def init
      product_versions = Hash.new { |hash, key| hash[key] = { channel: 'stable', version: 'latest' } }
      options[:product_versions].keys.each do |product|
        product_versions[product][:channel], product_versions[product][:version] = options[:product_versions][product].to_s.split(',', 2)
        product_versions[product][:version] ||= 'latest'
      end

      dev_lxc_template = %Q(<% unless options[:append] -%>
# enable_build_snapshots automatically makes container snapshots at key times during the build process
# default value is `true`
#enable_build_snapshots: true

# base_container must be the name of an existing container
base_container: <%= options[:base_container] %>

# memory_per_server sets the maximum amount of user memory (including file cache) for each server.
# dev-lxc will set the `memory.limit_in_bytes` cgroup for each server to apply this limit.
# If no units are specified, the value is interpreted as bytes.
# You can use suffixes to represent larger units - k or K for kilobytes, m or M for megabytes, and g or G for gigabytes.
# The default behavior is that no limit is set.
#memory_per_server: 4G

# list any host directories you want mounted into the servers
#mounts:
#  - /root/clusters root/clusters

# list any SSH public keys you want added to /home/dev-lxc/.ssh/authorized_keys
#ssh-keys:
#  - /root/clusters/id_rsa.pub

# DHCP reserved (static) IPs must be selected from the IP range 10.0.3.150 - 254
<% end -%>
<% if options[:chef] -%>

chef-server:
  users:          # a user's password will be the same as its username
    - mary-admin
    - joe-user
  orgs:
    demo:
      admins:
        - mary-admin
      non-admins:
        - joe-user
  servers:
    chef.lxc:
      ipaddress: 10.0.3.203
      products:
        <%- unless product_versions['chef-server'][:channel] == 'none' -%>
        chef-server:
          channel: <%= product_versions['chef-server'][:channel] %>
          version: <%= product_versions['chef-server'][:version] %>
        <%- end -%>
        <%- unless product_versions['manage'][:channel] == 'none' -%>
        manage:
          channel: <%= product_versions['manage'][:channel] %>
          version: <%= product_versions['manage'][:version] %>
        <%- end -%>
        <%- unless product_versions['push-jobs-server'][:channel] == 'none' -%>
        push-jobs-server:
          channel: <%= product_versions['push-jobs-server'][:channel] %>
          version: <%= product_versions['push-jobs-server'][:version] %>
        <%- end -%>
        <%- unless product_versions['reporting'][:channel] == 'none' -%>
        reporting:
          channel: <%= product_versions['reporting'][:channel] %>
          version: <%= product_versions['reporting'][:version] %>
        <%- end -%>
<% end -%>
<% if options[:chef_backend] -%>

chef-backend:
  api_fqdn: chef-ha.lxc
  users:          # a user's password will be the same as its username
    - mary-admin
    - joe-user
  orgs:
    demo:
      admins:
        - mary-admin
      non-admins:
        - joe-user
  servers:
    chef-backend1.lxc:
      ipaddress: 10.0.3.208
      role: backend
      leader: true
      products:
        <%- unless product_versions['chef-backend'][:channel] == 'none' -%>
        chef-backend:
          channel: <%= product_versions['chef-backend'][:channel] %>
          version: <%= product_versions['chef-backend'][:version] %>
        <%- end -%>
    chef-backend2.lxc:
      ipaddress: 10.0.3.209
      role: backend
      products:
        <%- unless product_versions['chef-backend'][:channel] == 'none' -%>
        chef-backend:
          channel: <%= product_versions['chef-backend'][:channel] %>
          version: <%= product_versions['chef-backend'][:version] %>
        <%- end -%>
    chef-backend3.lxc:
      ipaddress: 10.0.3.210
      role: backend
      products:
        <%- unless product_versions['chef-backend'][:channel] == 'none' -%>
        chef-backend:
          channel: <%= product_versions['chef-backend'][:channel] %>
          version: <%= product_versions['chef-backend'][:version] %>
        <%- end -%>
    chef-frontend1.lxc:
      ipaddress: 10.0.3.211
      role: frontend
      bootstrap: true
      products:
        <%- unless product_versions['chef-server'][:channel] == 'none' -%>
        chef-server:
          channel: <%= product_versions['chef-server'][:channel] %>
          version: <%= product_versions['chef-server'][:version] %>
        <%- end -%>
        <%- unless product_versions['manage'][:channel] == 'none' -%>
        manage:
          channel: <%= product_versions['manage'][:channel] %>
          version: <%= product_versions['manage'][:version] %>
        <%- end -%>
<% end -%>
<% if options[:chef_tier] -%>

chef-server:
  topology: tier
  api_fqdn: chef-tier.lxc
  users:          # a user's password will be the same as its username
    - mary-admin
    - joe-user
  orgs:
    demo:
      admins:
        - mary-admin
      non-admins:
        - joe-user
  servers:
    chef-be.lxc:
      ipaddress: 10.0.3.201
      role: backend
      bootstrap: true
      products:
        <%- unless product_versions['chef-server'][:channel] == 'none' -%>
        chef-server:
          channel: <%= product_versions['chef-server'][:channel] %>
          version: <%= product_versions['chef-server'][:version] %>
        <%- end -%>
        <%- unless product_versions['push-jobs-server'][:channel] == 'none' -%>
        push-jobs-server:
          channel: <%= product_versions['push-jobs-server'][:channel] %>
          version: <%= product_versions['push-jobs-server'][:version] %>
        <%- end -%>
        <%- unless product_versions['reporting'][:channel] == 'none' -%>
        reporting:
          channel: <%= product_versions['reporting'][:channel] %>
          version: <%= product_versions['reporting'][:version] %>
        <%- end -%>
    chef-fe1.lxc:
      ipaddress: 10.0.3.202
      role: frontend
      products:
        <%- unless product_versions['chef-server'][:channel] == 'none' -%>
        chef-server:
          channel: <%= product_versions['chef-server'][:channel] %>
          version: <%= product_versions['chef-server'][:version] %>
        <%- end -%>
        <%- unless product_versions['manage'][:channel] == 'none' -%>
        manage:
          channel: <%= product_versions['manage'][:channel] %>
          version: <%= product_versions['manage'][:version] %>
        <%- end -%>
        <%- unless product_versions['push-jobs-server'][:channel] == 'none' -%>
        push-jobs-server:
          channel: <%= product_versions['push-jobs-server'][:channel] %>
          version: <%= product_versions['push-jobs-server'][:version] %>
        <%- end -%>
        <%- unless product_versions['reporting'][:channel] == 'none' -%>
        reporting:
          channel: <%= product_versions['reporting'][:channel] %>
          version: <%= product_versions['reporting'][:version] %>
        <%- end -%>
<% end -%>
<% if options[:compliance] -%>

compliance:
  admin_user: admin         # the password will be the same as the username
  servers:
    compliance.lxc:
      ipaddress: 10.0.3.205
      products:
        <%- unless product_versions['compliance'][:channel] == 'none' -%>
        compliance:
          channel: <%= product_versions['compliance'][:channel] %>
          version: <%= product_versions['compliance'][:version] %>
        <%- end -%>
<% end -%>
<% if options[:supermarket] -%>

supermarket:
  servers:
    supermarket.lxc:
      ipaddress: 10.0.3.206
      products:
        <%- unless product_versions['supermarket'][:channel] == 'none' -%>
        supermarket:
          channel: <%= product_versions['supermarket'][:channel] %>
          version: <%= product_versions['supermarket'][:version] %>
        <%- end -%>
<% end -%>
<% if options[:automate] -%>

automate:
  servers:
    automate.lxc:
      ipaddress: 10.0.3.200
      products:
        <%- unless product_versions['automate'][:channel] == 'none' -%>
        automate:
          channel: <%= product_versions['automate'][:channel] %>
          version: <%= product_versions['automate'][:version] %>
        <%- end -%>
      license_path: ../delivery.license
      chef_org: automate
      enterprise_name: default
<% end -%>
<% if options[:build_nodes] -%>

build-nodes:
  servers:
    build-node-1.lxc:
      products:
        <%- unless product_versions['chefdk'][:channel] == 'none' -%>
        chefdk:     # downloaded only
          channel: <%= product_versions['chefdk'][:channel] %>
          version: <%= product_versions['chefdk'][:version] %>
        <%- end -%>
<% end -%>
<% if options[:runners] -%>

runners:
  servers:
    runner-1.lxc:
      products:
        <%- unless product_versions['chefdk'][:channel] == 'none' -%>
        chefdk:     # downloaded only
          channel: <%= product_versions['chefdk'][:channel] %>
          version: <%= product_versions['chefdk'][:version] %>
        <%- end -%>
<% end -%>
<% if options[:nodes] -%>

nodes:
  chef_server_url: https://chef.lxc/organizations/demo
  validation_client_name: demo-validator
  # comment out or remove the validation_key path to use chef-server keys generated by dev-lxc
  validation_key: # /path/for/ORG-validator.pem
  servers:
    node-1.lxc:
      products:
        <%- unless product_versions['chef'][:channel] == 'none' -%>
        chef:
          channel: <%= product_versions['chef'][:channel] %>
          version: <%= product_versions['chef'][:version] %>
        <%- end -%>
<% end -%>
<% if options[:analytics] -%>

analytics:
  servers:
    analytics.lxc:
      ipaddress: 10.0.3.204
      products:
        <%- unless product_versions['analytics'][:channel] == 'none' -%>
        analytics:
          channel: <%= product_versions['analytics'][:channel] %>
          version: <%= product_versions['analytics'][:version] %>
        <%- end -%>
<% end -%>
<% if options[:adhoc] -%>

adhoc:
  servers:
    adhoc.lxc:
      ipaddress: 10.0.3.207
<% end -%>
)

      dev_lxc_config = ERB.new(dev_lxc_template, nil, '-').result(binding)
      if options[:filename]
        mode = options[:append] ? 'a' : 'w'
        IO.write(options[:filename], dev_lxc_config, mode: mode)
      else
        puts dev_lxc_config
      end
    end

    desc "status [SERVER_NAME_REGEX]", "Show status of servers"
    option :config, :desc => "Specify a cluster's YAML config file. `./dev-lxc.yml` will be used by default"
    def status(server_name_regex=nil)
      cluster = get_cluster(options[:config])
      if cluster.config['chef-server'][:topology] == "tier" && cluster.config['chef-server'][:fqdn]
        printf "Chef Server FQDN: %s\n\n", cluster.config['chef-server'][:fqdn]
      end
      if cluster.config['chef-backend'][:fqdn]
        printf "Chef Server FQDN: %s\n\n", cluster.config['chef-backend'][:fqdn]
      end
      if cluster.config['analytics'][:topology] == "tier" && cluster.config['analytics'][:fqdn]
        printf "Analytics FQDN: %s\n\n", cluster.config['analytics'][:fqdn]
      end
      servers = Array.new
      cluster.get_sorted_servers(server_name_regex).map { |s| servers << s.status }
      max_server_name_length = servers.max_by { |s| s['name'].length }['name'].length unless servers.empty?
      servers.each_with_index do |s, server_index|
        printf "%-#{max_server_name_length}s     %-15s %s\n", s['name'], s['state'].upcase, s['ip_addresses']
        server = cluster.get_server(s['name'])
        server.snapshot_list.each do |snapname, snaptime, snap_comment|
          printf "  |_ %s %s %s\n", snapname, snaptime, snap_comment
        end
        puts if server_index + 1 < servers.length
      end
    end

    desc "attach [SERVER_NAME_REGEX]", "Attach the terminal to a single server"
    option :config, :desc => "Specify a cluster's YAML config file. `./dev-lxc.yml` will be used by default"
    def attach(server_name_regex)
      servers = get_cluster(options[:config]).get_sorted_servers(server_name_regex)
      if servers.length > 1
        puts "ERROR: The following servers matched '#{server_name_regex}'"
        servers.map { |s| puts "       #{s.name}" }
        puts "       Please specify a single server to attach to"
        exit 1
      elsif servers.empty?
        puts "ERROR: No servers matched '#{server_name_regex}'"
        puts "       Please specify a single server to attach to"
        exit 1
      end
      container = servers.first.container
      if !container.defined? || !container.running?
        puts "ERROR: Server '#{container.name}' is not running"
        exit 1
      end
      attach_opts = {
        wait: true,
        env_policy: LXC::LXC_ATTACH_CLEAR_ENV,
        extra_env_vars: ["LANG=en_US.UTF-8", "TERM=linux", "HOME=#{ENV['HOME']}"]
      }
      shell = ENV['SHELL']
      container.attach(attach_opts) { system(shell) }
    end

    desc "chef-repo", "Creates a '.chef' directory in the current directory using files from the cluster's backend /root/chef-repo/.chef"
    option :config, :desc => "Specify a cluster's YAML config file. `./dev-lxc.yml` will be used by default"
    option :force, :aliases => "-f", :type => :boolean, :desc => "Overwrite any existing knife.rb or pivotal.rb files"
    option :pivotal, :aliases => "-p", :type => :boolean, :desc => "Also copy pivotal.rb and pivotal.pem"
    def chef_repo
      get_cluster(options[:config]).chef_repo(options[:force], options[:pivotal])
    end

    desc "print-automate-credentials", "Print Automate credentials"
    option :config, :desc => "Specify a cluster's YAML config file. `./dev-lxc.yml` will be used by default"
    def print_automate_credentials
      get_cluster(options[:config]).print_automate_credentials
    end

    desc "run-command [SERVER_NAME_REGEX] [COMMAND]", "Runs a command in each server"
    option :config, :desc => "Specify a cluster's YAML config file. `./dev-lxc.yml` will be used by default"
    def run_command(server_name_regex=nil, command)
      start_time = Time.now
      get_cluster(options[:config]).get_sorted_servers(server_name_regex).each { |s| s.run_command(command); puts }
      print_elapsed_time(Time.now - start_time)
    end

    desc "prepare-product-cache [SERVER_NAME_REGEX]", "Download required product packages to cache"
    option :config, :desc => "Specify a cluster's YAML config file. `./dev-lxc.yml` will be used by default"
    def prepare_product_cache(server_name_regex=nil)
      start_time = Time.now
      cluster = get_cluster(options[:config])
      servers = cluster.get_sorted_servers(server_name_regex)
      cluster.prep_product_cache(servers, true)
      print_elapsed_time(Time.now - start_time)
    end

    desc "up [SERVER_NAME_REGEX]", "Start servers - This is the default if no subcommand is given"
    option :config, :desc => "Specify a cluster's YAML config file. `./dev-lxc.yml` will be used by default"
    def up(server_name_regex=nil)
      start_time = Time.now
      get_cluster(options[:config]).up(server_name_regex)
      print_elapsed_time(Time.now - start_time)
    end

    desc "halt [SERVER_NAME_REGEX]", "Shutdown servers"
    option :config, :desc => "Specify a cluster's YAML config file. `./dev-lxc.yml` will be used by default"
    def halt(server_name_regex=nil)
      start_time = Time.now
      get_cluster(options[:config]).halt(server_name_regex)
      print_elapsed_time(Time.now - start_time)
    end

    desc "snapshot [SERVER_NAME_REGEX]", "Manage a cluster's snapshots"
    option :comment, :aliases => "-c", :desc => "Add snapshot comment"
    option :config, :desc => "Specify a cluster's YAML config file. `./dev-lxc.yml` will be used by default"
    option :destroy, :aliases => "-d", :desc => "Destroy snapshot - use ALL to destroy all snapshots"
    option :list, :aliases => "-l", :type => :boolean, :desc => "List snapshots"
    option :restore, :aliases => "-r", :desc => "Restore snapshots"
    def snapshot(server_name_regex=nil)
      start_time = Time.now
      servers = get_cluster(options[:config]).get_sorted_servers(server_name_regex)
      if options[:list]
        servers.each_with_index do |s, server_index|
          puts s.name
          s.snapshot_list.each do |snapname, snaptime, snap_comment|
            printf "  |_ %s %s %s\n", snapname, snaptime, snap_comment
          end
          puts if server_index + 1 < servers.length
        end
        return
      elsif options[:destroy]
        snapname = options[:destroy] == 'destroy' ? "LAST" : options[:destroy]
        servers.each { |s| s.snapshot_destroy(snapname); puts }
      elsif options[:restore]
        running_servers = Array.new
        servers.each do |s|
          running_servers << s.name if s.container.running?
        end
        unless running_servers.empty?
          puts "ERROR: Aborting snapshot restore because the following servers are running"
          puts running_servers
          exit 1
        end
        snapname = options[:restore] == 'restore' ? "LAST" : options[:restore]
        servers.each { |s| s.snapshot_restore(snapname); puts }
      else
        running_servers = Array.new
        servers.each do |s|
          running_servers << s.name if s.container.running?
        end
        unless running_servers.empty?
          puts "ERROR: Aborting snapshot because the following servers are running"
          puts running_servers
          exit 1
        end
        servers.each { |s| s.snapshot(options[:comment]); puts }
      end
      print_elapsed_time(Time.now - start_time)
    end

    desc "destroy [SERVER_NAME_REGEX]", "Destroy servers"
    option :config, :desc => "Specify a cluster's YAML config file. `./dev-lxc.yml` will be used by default"
    option :force, :aliases => "-f", :type => :boolean, :desc => "Destroy servers without confirmation"
    def destroy(server_name_regex=nil)
      servers = get_cluster(options[:config]).get_sorted_servers(server_name_regex)
      if servers.empty?
        puts "No matching server names were found"
        exit
      end
      unless options[:force]
        confirmation_string = String.new
        servers.reverse_each { |s| confirmation_string += "#{s.name}\n" }
        confirmation_string += "Are you sure you want to destroy these servers? (y/N)\n"
        return unless yes?(confirmation_string)
      end
      start_time = Time.now
      get_cluster(options[:config]).destroy(server_name_regex)
      print_elapsed_time(Time.now - start_time)
    end

  end
end
